Source/Core/InputCommon/DynamicInputTextureConfiguration.cpp

```diff
@@ -0,0 +1,296 @@
+      return;
+    }
+
+    for (auto& [host_control, image_name] : host_controls)
+    {
+      texture_data.m_host_input_to_input_image.insert_or_assign(host_control, image_name.to_str());
+    }
+
+    // Default to the default device but overwrite if the creator
+    // has provided something specific
+    texture_data.m_devicename = default_device;
+    const picojson::value& device_json = data.get("device");
+    if (device_json.is<std::string>())
+    {
+      texture_data.m_devicename = device_json.get<std::string>();
+    }
+
+    if (texture_data.m_devicename.empty())
+    {
+      ERROR_LOG(VIDEO,
+                "Failed to load dynamic input json file '%s' because field "
+                "'device' is missing ",
+                json_file.c_str());
+      m_valid = false;
+      return;
+    }
+
+    m_dynamic_input_textures.emplace_back(std::move(texture_data));
+  }
+}
+
+DynamicInputTextureConfiguration::~DynamicInputTextureConfiguration() = default;
+
+void DynamicInputTextureConfiguration::GenerateTextures(const IniFile::Section* sec) const
+{
+  bool any_dirty = false;
+  for (const auto& texture_data : m_dynamic_input_textures)
+  {
+    any_dirty |= GenerateTexture(sec, texture_data);
+  }
+
+  if (!any_dirty)
+  {
+    return;
+  }
+  g_renderer->ForceReloadTextures();
+}
+
+bool DynamicInputTextureConfiguration::GenerateTexture(
+    const IniFile::Section* sec, const DynamicInputTextureData& texture_data) const
+{
+  std::string device_name;
+  if (!sec->Get("Device", &device_name) || device_name != texture_data.m_devicename)
+  {
+    return false;
+  }
+
+  // Load colorkey image copy
+  auto color_key_image = LoadImage(m_base_path + texture_data.m_colorKeyImageName);
+  bool dirty = false;
+  for (auto& [emulated_key, color] : texture_data.m_emulated_input_to_color)
+  {
+    std::string host_key = "";
+    sec->Get(emulated_key, &host_key);
+
+    const auto color_only_mask = MaskColor(*color_key_image, Pixel{color.r, color.g, color.b, 255});
+    const auto rects = FindRectRegionsOfNeighboringPixels(color_only_mask);
+
+    const auto input_image_iter = texture_data.m_host_input_to_input_image.find(host_key);
+    if (input_image_iter == texture_data.m_host_input_to_input_image.end())
+    {
+      const auto fallback_image = LoadImage(m_base_path + texture_data.m_fallbackImageName);
+
+      for (const auto& rect : rects)
+      {
+        CopyTo(*fallback_image, *color_key_image, rect, rect);
+        dirty = true;
+      }
+    }
+    else
+    {
+      const auto host_key_image = LoadImage(m_base_path + input_image_iter->second);
+
+      for (const auto& rect : rects)
+      {
+        if (host_key_image->width == rect.width && host_key_image->height == rect.height)
+        {
+          CopyTo(*host_key_image, *color_key_image, Rect{Point{0, 0}, rect.width, rect.height},
+                 rect);
+          dirty = true;
+        }
+        else
+        {
+          if (texture_data.m_preserve_aspect_ratio)
```

Comment:
Is there any particular reason for this `else` not being joined with the `if` into an `else if`? There is no code directly inside this outer `else` (unless there should be, and it is missing?)

Reply:
Nope, just a style thing.  I fixed it!